### ISSUE TITLE
BUG: users with a space in the name cannot be at-mentioned

### ISSUE THREAD

**Issue by ryjones** [RocketChat/Rocket.Chat#10253]
### Description:

We allow the use of social logins. Users may come in from GitHub, or Google+, for instance and put a space in the username when they sign up. I am unable to at-mention these users.

### Server Setup Information:

* Version of Rocket.Chat Server: 0.62.0

### Steps to Reproduce:

1. create user with space in name
2. join user to channel
3. attempt to at-mention user
4. User appears in pick list, is chosen

### Expected behavior:

5. alert is sent to user with space in name

### Actual behavior:

5. no alert is sent

**Comment by ryjones** [comment]
@milesflo valid usernames may have spaces, though

**Comment by milesflo** [comment]
@ryjones No, when I say trim, I mean removing from either end of the string, not filtering whitespace out entirely... Like literally just call [`trim()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim) as the first step of validating a username before setting it.

**Comment by ChameleonScales** [comment]
this issue is also discussed in #956, although it deserves its own thread (here).

**Comment by Ablesius** [comment]
> @ryjones No, when I say trim, I mean removing from either end of the string, not filtering whitespace out entirely... Like literally just call [`trim()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim) as the first step of validating a username before setting it.

How would this apply to white space in the middle of a username, though?

**Comment by milesflo** [comment]
> > @ryjones No, when I say trim, I mean removing from either end of the string, not filtering whitespace out entirely... Like literally just call [`trim()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim) as the first step of validating a username before setting it.
> 
> How would this apply to white space in the middle of a username, though?

This post is over a year old 🙂 DW about it

**Comment by Ablesius** [comment]
Hope you don't mean to say this issue is not going to be resolved. It's been a PITA for at least 2 years.

**Comment by ryjones** [comment]
This is still a huge problem for Hyperledger; I imagine it impacts other instances as well.

**Comment by Ablesius** [comment]
It is still an issue for me. Is there a follow-up issue or can we please re-open this one?

**Comment by ChameleonScales** [comment]
@ryjones You may have authored this issue but since several other people have shown interest in it, not caring anymore yourself is no longer a good enough reason to close it.

**Comment by Karreg** [comment]
Any news on the subject? We just hit this issue. In France there are a lot of people with spaces in their names... :(

**Comment by Ablesius** [comment]
After reading the thread on the special characters in the other issue, I get that this problem is a bit bigger, but still, a single whitespace in between two words like "First Last" should work imho.

**Comment by ryjones** [comment]
@Sing-Li @geekgonecrazy this is still a huge problem for Hyperledger.

**Comment by eig0r** [comment]
any update for this issue ?

**Comment by Ablesius** [comment]
Is anybody even looking into this? Are there some hints for people willing to contribute as to where to even start looking? This issues is still a huge problem for us and we're actually considering dropping rocket.chat. I'd be willing to try and contribute but I'd need some pointers.

**Comment by ryjones** [comment]
@Ablesius we're also looking for rocket.chat replacements

**Comment by geekgonecrazy** [comment]
Feels like we should send the mention along with the message instead of purely parsing the sent message server side for a mention.  Especially cases like this where client literally auto completes.  We wouldn’t have to take client word for a mention if we didn’t want to and scan the message for the mention.

**Comment by jeremyevans6** [comment]
In my case this is due to a lack of alternatives when creating accounts via Oauth. I don't want to modify the package (I'm on Snap even), so I think I'll be running a DB find and replace to correct this and maybe making it a cronjob. Alternatively, and better, I could put in a hook upon account creation. 

Does anyone have experience with any patches like this?

**Comment by jeremyevans6** [comment]
Here's my patch to the DB in a snap instance:

`
#
#  BAND AID for Username with spaces
#
# sudo rocketchat-server.mongo
# ******************
use parties
db.users.find({username: { $regex: /.*\s.*/ }}).forEach(function(e,i) {    e.username=e.username.replace(" ","");    db.users.save(e);});
db.users.find({ "services.tsol.username": { $regex: /.*\s.*/ }}).forEach(function(e,i) {    e.services.tsol.username=e.services.tsol.username.replace(" ","");    db.users.save(e);});
# ******************
`

Obviously, this is something that needs to be run regularly to prevent this problem. Because I'm dealing with a closed community which receives members only occassionally, this is fine for me.

Maybe someone knows how to make a mongo shell command execute as a cronjob?

**Comment by mik2** [comment]
Maybe there should be another way of doing mention (besides the current one), which would help in confusing situations, e.g. `@{username}`, where `{` and `}` are characters that cannot appear in username.
